### ISSUE TITLE
Use Rust union types

### ISSUE THREAD

**Issue by asomers** [rust-lang/libc#1020]
Rust gained a union type in [`1.19.0`](https://blog.rust-lang.org/2017/07/20/Rust-1.19.html).  libc wraps many C unions poorly, usually by pretending they're structs and ignoring some fields.  For example, https://github.com/rust-lang/libc/blob/master/src/unix/mod.rs#L156 .  libc should update all of these to use the correct union types before release 1.0.  It should be fairly easy to find them all, because all of them are listed as exclusions in libc-test/build.rs.

**Comment by faern** [comment]
I guess we can replace existing structs with real unions. Just that it would be a breaking change, needing a major version bump?

I hope all types that represent unions are going to be implemented as unions before `libc 1.0`? The current representations feels like hacks that were basically just waiting for Rust 1.19?

**Comment by alexcrichton** [comment]
@faern yes new types can probably use `union` as necessary, and if/when libc 1.0 happens we can switch to unions for existing types.

**Comment by tgross35** [comment]
We've been using unions for a while, and the needed breaking changes have at least made it to `main`.
